### DISCUSSION TITLE
The path name of the pocketbase executable is tied to the database

### DISCUSSION THREAD

**Discussion by IwateKyle** [pocketbase/pocketbase#2143] [Q&A]
The problem below seems to be resolved.  But I really struggled with this for about 30 minutes.  Then all of the sudden, I was getting the expected results.  I'll leave this discussion here in case anyone else was having the same experience.  Otherwise if someone wants to close it - okay.

===============================================================================
Is there a reason why the database is directly tied to the user facing executable name.

Where pb_data is the same location, all the below try to create a new database upon first access.
* pocketbase serve
* ~/bin/pocketbase serve
* ~/bin/myapp serve

This makes it difficult to
* move to production where the path is not necessarily the same as development
* change the name of the executable after database has been created
* move the executable to a new location without jumping through hoops using aliases and softlinks.

It seems that if there were some reason to have multiple databases created in the same pb_data folder, then a command line option should be used to address which database is intended to be accessed, not the name of the executable.

Is there a work-around.  I have already created data, but the path will not work for me in production.  How can I change the hard linked path in the database?

**Comment by ganigeorgiev** [comment]
I don't understand what do you mean. The executable name has nothing to do with the database name.

You can rename the executable however you want.

`pb_data` by default will be created next to the executable. If you want to specify a different `pb_data` location, you can make use of the `--dir` flag:


```sh
./pocketbase serve --dir="/path/to/your/pb_data_dir"
```

You can list all available flags with `--help`.

**Reply by IwateKyle** [reply]
After I created my first DB, I copied the data folder to a backup folder because I already had collections with data.  I wanted to see what would happen when I created my own app using Pocketbase as a framework.  Originally, I compiled my app with basic defaults, so the program name was initially just main.

from within the backup folder which was next to my data folder, I ran

`../app/main serve
`
It first appeared to work fine, but when I looked closer, it was prompting me to create a new admin account.

Next, I tried changing the name of my compiled app to pocketbase and ran
`../app/pocketbase serve`

again same problem.
next, I exported the app folder
```
export PATH=/home/..../app:$PATH
which pocketbase
=> /home/..../app/pocketbase
```
I ran pocketbase serve
`pocketbase serve`
and I was able to log in just fine.
I made sure to run `which` to make sure that I was using the new executable that I compiled.
anyway - that worked.  
My new executable with the same name as the Pocketbase executable was the only connection I was able to make where both were called with out using a hard coded path, or relative path, or a different name.
That's when I changed the name of the program and it still broke again.

I was struggling for a long time.  

After I posted this message, I continued having the same problem, but then all of the sudden, the problem went away.  I have no idea what was causing the phenomenon.

As I said, you can close this if you want.  I just left it open in case anyone else was experiencing the same issue.

**Reply by ganigeorgiev** [reply]
I'm sorry but I still don't understand what you mean.

The database filename has nothing todo with the executable filename.

By default `pb_data` will be created next to the executable. Aka. if your executable is located in `/home/dev/app`, then the `pb_data` will be created there.

As mentioned in the comment above, if you want to specify an explicit `pb_data` location, you can do that with `--dir` flag:


```sh
./pocketbase serve --dir="/path/to/your/pb_data_dir"
```

**Reply by nicosql** [reply]
@IwateKyle I had similar struggles until I read the docs thoroughly, hopefully things are smoother for you now.  I'm really enjoying this package.  Although I'm certain your issue was just a lack of understanding I wanted to thank you for posting this because it still prompted @ganigeorgiev to answer questions I had about renames.

Good stuff guys, thanks @ganigeorgiev for this great work - keep going!


--a guy who loves pocketbase and sends @ganigeorgiev money 😅

**Reply by IwateKyle** [reply]
@pookiepats , thanks for chiming in.  Now I don't feel like I was going absolutely crazy.  Phew.  Yes, I agree, it's a wonderful tool, especially for single operators like myself.

**Reply by mdnahas** [reply]
I have a similar question/issue.  pocketbase read/writes its files in relationship to the location of the binary, not the current working directory.

When I run "../pocketbase serve", it creates directories "pb_data" and "pb_migrations" in the parent directory (where the binary is located) and not in the current directory.

That feels wrong.

**Reply by ganigeorgiev** [reply]
@mdnahas You can explicitly set custom `pb_data` and `pb_migrations` directories with the `--dir` and `--migrationsDir` flags to accommodate your use case.

**Reply by mdnahas** [reply]
@ganigeorgiev Yes, I know that.  I'm saying that almost every other program operates in the directory where they are run, not the location of the binary.  PocketBase's current behavior is the unexpected.  It is breaking convention.  You should only break convention if you think there's a huge win, and I don't think see it here.  

Right now, I'm learning pocketbase and making a lot of demo directories.  I don't want to copy the binary into each one.  I don't want to do the lengthy "--dir demo4/pb_data --migrationDir demo4/pb_migrations ..." every time I invoke pocketbase.  

My current solution is to do symbolic links to the binary in each demo directory.  That seems to work.  I hate using symbolic links this way, but it works.  

But more than anything, pocketbase is breaking convention.  Which means its users will keep stumbling into this problem, like I did.  **Please have it match the convention** (or give me a very good reason for it to break the convention!).

**Reply by ganigeorgiev** [reply]
@mdnahas I don't understand what do you refer by convention. It is just an implementation detail.

The main idea behind the current defaults is that the binary usually should stay together with your project files (or at least that's how I'm planning to use it). You can have multiple projects and each can use different PocketBase version and/or configurations.

PocketBase is distributed as a regular Go package and you can use it as a library as part of your own Go program. The current prebuilt executable is just a simple example created from the [examples/base/main.go](https://github.com/pocketbase/pocketbase/blob/master/examples/base/main.go) file. You can always create your own if you don't like the current defaults.

At the end of the day, this is just a small hobby open source project that I've created primarily for my own needs (see [Presentator #183](https://github.com/presentator/presentator/issues/183)) and I'm not trying to sell you anything. If you find it useful - great! If not - there are plenty of other more mature solutions maintained by larger teams that can provide better support and more features (just to mention some - Supabase, Nhost, Appwrite, SurrealDB, EdgeDB, Strapi, Directus, etc.)

**Reply by mdnahas** [reply]
@ganigeorgiev By "convention", I mean "every other program runs this way".  E.g., "ls", "cat", etc..  It is easier to learn a program if it behaves like every other program.  

It is a great project!  I'm coding a small site for a friend's small business.  I may not be around to maintain it, so I wanted something simple that he could understand and keep working.  And now, having used PocketBase a little, I plan on using it a lot.  I've been programming my whole life and I've had an iPhone for 13 years that I've never programmed for --- I can easily whip up a fun program for it in a few days with PocketBase!

BTW, I found out about it on HackerNews.  When someone asked what backend to use, someone said it was the only one that made backends fun to program.  I understand why they said that!  

It's your own project, so do what you like.  But you can expect other users to run into this problem, because it diverges from the convention.
